lib/rpush/client/active_model/fcm/notification.rb

```diff
@@ -0,0 +1,121 @@
+
+          ROOT_NOTIFICATION_KEYS = %w[title body image].freeze
+          ANDROID_NOTIFICATION_KEYS = %w[icon tag color click_action body_loc_key body_loc_args title_loc_key
+                                         title_loc_args channel_id ticker sticky event_time local_only
+                                         default_vibrate_timings default_light_settings vibrate_timings
+                                         visibility notification_count light_settings sound].freeze
+
+          def self.included(base)
+            base.instance_eval do
+              validates :device_token, presence: true
+              validates :priority, inclusion: { in: FCM_PRIORITIES }, allow_nil: true
+
+              validates_with Rpush::Client::ActiveModel::PayloadDataSizeValidator, limit: 4096
+              validates_with Rpush::Client::ActiveModel::RegistrationIdsCountValidator, limit: 1000
+
+              validates_with Rpush::Client::ActiveModel::Fcm::ExpiryCollapseKeyMutualInclusionValidator
+              validates_with Rpush::Client::ActiveModel::Fcm::NotificationKeysInAllowedListValidator
+            end
+          end
+
+          def payload_data_size
+            multi_json_dump(as_json['message']['data']).bytesize
+          end
+
+          # This is a hack. The schema defines `priority` to be an integer, but FCM expects a string.
+          # But for users of rpush to have an API they might expect (setting priority to `high`, not 10)
+          # we do a little conversion here.
+          def priority=(priority)
+            case priority
+            when 'high', FCM_PRIORITY_HIGH
+              super(FCM_PRIORITY_HIGH)
+            when 'normal', FCM_PRIORITY_NORMAL
+              super(FCM_PRIORITY_NORMAL)
+            else
+              errors.add(:priority, 'must be one of either "normal" or "high"')
+            end
+          end
+
+          def dry_run=(value)
+            fail ArgumentError, 'FCM does not support dry run' if value
+          end
+
+          def mutable_content=(value)
+            fail ArgumentError, 'RPush does not currently support mutable_content for FCM' if value
+          end
+
+          def as_json(options = nil) # rubocop:disable Metrics/PerceivedComplexity
+            json = {
+              'data' => data,
+              'android' => android_config,
+              'token' => device_token
+            }
+            # Android does not appear to handle content_available anymore. Instead "priority" should be used
+            # with "low" being a background only message. APN however should support this field.
+            # json['content_available'] = content_available if content_available
+            json['notification'] = root_notification if notification
+            { 'message' => json }
+          end
+
+          def android_config
+            json = {
+              'notification' => android_notification,
+            }
+            json['data'] = data if data
```

Comment:
Does this need to be duplicated here? I think this is only for Android-specific override if necessary and since we are not using FCM to deliver to both platforms, not having to duplicate this key makes the payload smaller

I see similar duplicated values like `notification_priority` within the Android overrides. In general, I think the first phase of FCM we are adding is only to deliver to Android, Apns8 will continue to be used for iOS. So we can avoid duplicating platform-overrides for now.

How we use rpush will need to change if we are to deliver to both iOS and Android via FCM and that's not the intention right now anyway.